### PR TITLE
cocotb package

### PR DESCRIPTION
**Pull request by themperek**
Simple setup.py that allow packaging cocotb.
It allows installing with `pip install cocotb==1.0.dev3`. 

- change `build` folder to a local directory (no need to have access to installation directory) 
- one can get the path to` COCOTB` with `$(shell cocotb-config --share)` (no need to set `COCOTB` environment variable) 
- remove `SPHINX_BUILD` environmental variable and replace with `COCOTB_SIM` to recognize if loading simulation. This makes easy multi-platform support. 
- using tox to run the tests 

I have modified examples but not tests.
After merge will automatically package build.
It should be backward compatible (symlink to makefile folder)

### CODE REVIEW

**Comment by themperek** [comment]
One improvement would be to precompile the libraries during installation. For windows distribute precompiled libraries. This will work if no customization is need during simulation later?

**Comment by themperek** [comment]
Windows: cocotb-path will not work on windows out of box. This can be done bit different but SPHINX_BUILD variable has to go away. As far as I can see one would need to be able `import cocotb` without error without SPHINX_BUILD.

**Comment by stuarthodgson** [comment]
@themperek This stands a good chance of getting pulled in as is after I have tested in our development environment. As you also state I wholly agree about pre compiling binaries as well during "instalation" but pip is really no good at that. Maintaining pre-compiled libs is also not a path I have much interest in going down given how much that can bolloon (also do not use windows). Hope to pull this in relatively soon or at least make some suggestions.

**Comment by themperek** [comment]
I can work on this and make it Windows compatible but would need to have the possibility to `import cocotb` without setting any environmental variable (SPHINX_BUILD). 

This can be important change but can we recognise the simulation based on `COCOTB_SIM` variable (if set then simulation) and remove SPHINX_BUILD variable?

**Comment by themperek** [comment]
Should now work with:  `pip install cocotb`

**Comment by forensicgarlic** [comment]
I'm a little unclear from reading the updates, this works fine for linux, but might have windows issues?

**Comment by elmsfu** [comment]
@themperek I was able to remove the need for SPHINX_BUILD [here](https://github.com/elmsfu/cocotb/commit/59579cf4758aeeebef8899066b8d65d679d06d8b). Although I realized that by moving where the makefiles, lib, and include are installed (under site-packages/cocotb where I feel they should be), that they are now inconsistent with the structure of the repository.

@stuarthodgson Would you be open to moving the support files under the cocotb module as they are all in support of that python module?

**Comment by TC01** [comment]
If nothing else, on multi-user systems it makes it really easy to ensure everyone is using the same version of cocotb. (Even without installing through pip, moving the ```build``` folder to the local directory makes it far easier to have one copy/specific version of cocotb used by multiple people).

**Comment by elmsfu** [comment]
pip also helps with python dependencies, but I don't that applies to cocotb at the moment.

cocotb is unique in how python is invoked and interacts with the simulators. I think it would definitely be worth while if a move toward a more pythonic approach (e.g. #437).

**Comment by themperek** [comment]
>I don't have a problem with it. Can someone explain to me what we get by having cocotb pip installable but still needed to have the libs compiled?

- so that you do not have to set any extra environmental variable or download anything
- so it is easy to install/update/unistall 
- so cocotb is easy to add as requirements for other packages

Next step for me would be to precompile the libraries during installation so one could provide packages  without need of compilers etc.

File: cocotb/path.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+
+import os
+import imp
```
Comment (eric-wieser) [inline cocotb/path.py]: This is deprecated as of python 3.4 - you should do use `importlib.find_spec('cocotb').origin` there

Comment (eric-wieser) [reply cocotb/path.py]: What are you gaining by doing this instead of using `cocotb.__file__`?

**Comment by imphil** [comment]
This looks like a good step forward, thanks for the effort @themperek.

However, I think it's crucial to add the compilation of the libraries to the packaging as well. Today the compilation step is a major pain point (in my experience), and it becomes even more so if users use "sudo pip install cocotb" and then try to run cocotb as normal user, in which case they cannot compile the libraries as they don't have (or should have) access to the installation directory.
Including the compilation step does *not* mean shipping pre-compiled libraries, this is orthogonal and optional.

But, since perfect is the enemy of good, I'm open to take this PR as-is, and improve the situation in further PRs later on.

@themperek Before merging can you clean up the commit history a bit using `git rebase -i`?

**Comment by TC01** [comment]
> However, I think it's crucial to add the compilation of the libraries to the packaging as well. Today the compilation step is a major pain point (in my experience), and it becomes even more so if users use "sudo pip install cocotb" and then try to run cocotb as normal user, in which case they cannot compile the libraries as they don't have (or should have) access to the installation directory.

Part of this pull request is to make it so that the libraries are compiled locally (in your current working directory) instead of in the cocotb installation directory-- a local "build" folder gets created when running. So this is something that's already fixed-- unless I'm missing something?

One other thing, though, is that this appears to pollute the ```python/site-packages``` directory:

```
$ ls ~/.local/lib/python2.7/site-packages/ -l
total 132
drwxrwxr-x.  5 bjr bjr  4096 Nov 14 14:52 cocotb
drwxrwxr-x.  2 bjr bjr  4096 Nov 14 14:52 cocotb-1.0.1-py2.7.egg-info
drwxrwxr-x.  2 bjr bjr  4096 Nov 14 14:52 include
drwxrwxr-x. 11 bjr bjr  4096 Nov 14 14:52 lib
drwxrwxr-x.  3 bjr bjr  4096 Nov 14 14:52 makefiles
drwxrwxr-x.  2 bjr bjr  4096 Sep 17 15:04 tests
```

If it's installed in a virtualenv this isn't such a problem, but it would be really nice to install the include, lib, and makefiles folders below "cocotb" rather than into site-packages directly.

**Comment by themperek** [comment]
@imphil 
>However, I think it's crucial to add the compilation of the libraries to the packaging as well. 

It is compiled locally now.  If you mean to compile during setup then this is not so easy. As I understand libraries are specific to the simulator and interface. One could precompile for all available simulators but not sure if this is much better. Distributing binaries, in this case, would be hard.

**Comment by themperek** [comment]
@TC01 

>One other thing, though, is that this appears to pollute the python/site-packages directory.

One would need to restructure the repository or make some coping during setup (this will most likely break backward compatibility in both cases). You can make an new issue.

**Comment by themperek** [comment]
One can try now `pip install cocotb==1.0.dev1`

Descryption: https://pypi.org/project/cocotb/1.0.dev1/

Will marge next days if no comments.

**Comment by stuarthodgson** [comment]
I'd just like to try before merging. Also I think we already have 1.0 as a tag but that won't match this 1.0 soaybe we merge, tag, then do what is needed on pypi at 1.1.

**Comment by themperek** [comment]
Please wait a day.

**Comment by imphil** [comment]
> @imphil
> 
> > However, I think it's crucial to add the compilation of the libraries to the packaging as well.
> 
> It is compiled locally now. If you mean to compile during setup then this is not so easy. As I understand libraries are specific to the simulator and interface. One could precompile for all available simulators but not sure if this is much better. Distributing binaries, in this case, would be hard.

Ah ok, I misread the code when skimming over it. Sounds good!

**Comment by imphil** [comment]
Regarding the versioning thing: even though it's not fantastic, setuptools_scm works reasonable well (https://github.com/pypa/setuptools_scm) and makes sure the SCM version is equal to the package version, even in local and development builds.

**Comment by themperek** [comment]
>I'd just like to try before merging

@stuarthodgson Any news?

**Comment by stuarthodgson** [comment]
Some but only on a related note of syncing our internal fork with upstream in preparation for a big or so we can drop out internal fork and move to upstream again. Which will mean it gets more focus. I'll try tomorrow.to test this

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+author_email = 'cocotb@potentialventures.com'
+
+install_requires = []
+
+def package_files(directory):
+    paths = []
+    for (fpath, directories, filenames) in walk(directory):
+        for filename in filenames:
+            paths.append(path.join('..', fpath, filename))
+    return paths
```
Comment (imphil) [inline setup.py]: This functions breaks the directory layout in the packaging by having the `extra_files` include a `..` at the beginning of each path. In the end `package_data` is installed into `site-packages` directly and not into the `cocotb` subdirectory where it belongs.

```sh
$ ll lib/python3.6/site-packages/
insgesamt 16
drwxr-xr-x  6 philipp users 4096  7. Dez 11:08 cocotb
drwxr-xr-x  2 philipp users  141  7. Dez 11:08 cocotb-1.0.1-py3.6.egg-info
-rw-r--r--  1 philipp users  126  7. Dez 11:07 easy_install.py
drwxr-xr-x  2 philipp users  135  7. Dez 11:08 include
drwxr-xr-x 11 philipp users  139  7. Dez 11:08 lib
drwxr-xr-x  3 philipp users  234  7. Dez 11:08 makefiles
drwxr-xr-x 11 philipp users 4096  7. Dez 11:07 pip
drwxr-xr-x  2 philipp users  155  7. Dez 11:07 pip-9.0.3.dist-info
drwxr-xr-x  5 philipp users   94  7. Dez 11:07 pkg_resources
drwxr-xr-x  2 philipp users   41  7. Dez 11:07 __pycache__
drwxr-xr-x  6 philipp users 4096  7. Dez 11:07 setuptools
drwxr-xr-x  2 philipp users  218  7. Dez 11:07 setuptools-39.0.1.dist-info
```

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    long_description=long_description,
+    long_description_content_type='text/markdown',
+    author=author,
+    maintainer=author,
+    author_email=author_email,
+    maintainer_email=author_email,
+    install_requires=install_requires,
+    packages=find_packages(),
+    include_package_data=True,
+    package_data={'cocotb': extra_files},
```
Comment (imphil) [inline setup.py]: How about having the function `package_files()` take multiple directories, and call it directly here? (In line with the other packaging information going directly into this dict as opposed to intermediate variables.)

File: MANIFEST.in [hunk 4]
```diff
@@ -5,6 +5,8 @@ recursive-include vpi_shim *.c Makefile
 recursive-include examples *.v *.py Makefile *.tcl *.cfg
 recursive-include scripts *
 recursive-include simulator *.c *.h Makefile
+recursive-include lib *.c* *.h Makefile
+recursive-include include *.h
 recursive-include modules *.py
 recursive-include makefiles Makefile.*
-include Makefile README.md
+include Makefile README.md version
```
Comment (imphil) [inline MANIFEST.in]: The `MANIFEST.in` file only needs to list files which appear in the source package (sdist), but not in the binary package (bdist). That is, all files which are already included in some way in the setup.py file, e.g. in  `package_data` should not be added here.

This leaves most likely only the examples.

File: setup.py (hunk 3 above)
Comment (imphil) [inline setup.py]: Another point on `package_data`: `Makefile.sim` requires the `version` file to be present:

```sh
include $(SIM_ROOT)/version
export VERSION
```

Therefore the version file needs to be included in `package_data` and installed as part of the package.

File: setup.py (hunk 2 above)
Comment (themperek) [reply setup.py]: See: https://github.com/potentialventures/cocotb/pull/517#issuecomment-438804015 
I agree this is not nice. The only easy way I out see is to move `makefiles`, `lib` and `include` into a cocotb folder (maybe `cocotb/src`). This will be a major change to the project structure. Was planning to do this in a sperate PR after some discussion. Can change now. Suggestions welcome.

Comment (imphil) [reply setup.py]: Ah, I missed that comment. To me it's not only "not nice;" but a showstopper -- just install https://pypi.org/project/include/ for example and you have a naming clash in the site-packages directory. This leaves two options:

1. Change the source code layout to move all installed components into the cocotb directory. This would be typical approach for most Python packages I've seen. However, there's a huge risk of breaking the assumptions that some people might have regarding the source code structure (since that's how cocotb is used these days). So that's a dangerous option.

2. Only restructure the layout within the package. I didn't look closely how this could be accomplished with setuptools, but there must be a way? This would enable us to change the layout in the package only, and if that turns out to be successful, we can do the same to the repo layout. If that's doable (even if it's painful) I'd take that option.

Could you look into the feasibility of these options?

Comment (imphil) [reply setup.py]: A quick google search reveals a couple options. Using `data_files` instead of `package_data` is one, https://stackoverflow.com/a/26533921 shows another one (it's hacky but should do the trick, at least on Linux)

Comment (themperek) [reply setup.py]: I already tried "few" solutions but I can be missing something.

- https://stackoverflow.com/a/26533921 : `os.symlink` will not work always on Windows but can try to use `shutil.copy` to `cocotb/src` (maybe only for `install` command).
- `data_files` will be installed into `sys.prefix` directory. If this is OK I can do. There is a workaround but it does not work for me: https://groups.google.com/forum/#!topic/comp.lang.python/Nex7L-026uw

Any suggestions?

Comment (imphil) [reply setup.py]: Reading through the PR again I see that @stuarthodgson mentioned in https://github.com/potentialventures/cocotb/pull/517#issuecomment-385794426 that he'd be OK with moving the files into the `cocotb` folder. Given that and the trouble with doing it otherwise we probably should just go with that. How about introducing a `share` directory (not entirely happy with the name, probably the perfect choice doesn't exist)

```
makefiles -> cocotb/share/makefiles
include -> cocotb/share/includes
lib -> cocotb/share/lib
```

I tried doing that quickly as a separate patch but it becomes absolutely messy with all the tests, so it's best done together with the changes in this PR.

**Comment by imphil** [comment]
While thinking about the directory structure I realized that we can actually follow a bit more established conventions by renaming "cocotb-path" to "cocotb-config", and giving it command arguments like "--makefiles" to get the path to the makefiles, or `cocotb-config --libdir` to get the libdir, or "--incdir" to get the include directory. This is in line of how tools used to do it before "pkg-config" provided an universal solution for that. For example, ncurses6-config outputs  that on my machine:

```
$ ncurses6-config --help
Usage: ncurses6-config [options]

Options:
  --prefix           echos the package-prefix of ncurses
  --exec-prefix      echos the executable-prefix of ncurses

  --cflags           echos the C compiler flags needed to compile with ncurses
  --libs             echos the libraries needed to link with ncurses

  --version          echos the release+patchdate version of ncurses
  --abi-version      echos the ABI version of ncurses
  --mouse-version    echos the mouse-interface version of ncurses

  --bindir           echos the directory containing ncurses programs
  --datadir          echos the directory containing ncurses data
  --includedir       echos the directory containing ncurses header files
  --libdir           echos the directory containing ncurses libraries
  --mandir           echos the directory containing ncurses manpages
  --terminfo         echos the $TERMINFO terminfo database path
  --terminfo-dirs    echos the $TERMINFO_DIRS directory list
  --termpath         echos the $TERMPATH termcap list

  --help             prints this message
```

File: setup.py (hunk 2 above)
Comment (themperek) [reply setup.py]: This would be easiest.  I can work on this.
Maybe should keep symbolic links to those folders for some time in the top directory and issue a warning that COCOTB envirmant variable is obsolete?

Comment (themperek) [reply setup.py]: Most of the existing Makefiles have to be changed.  I expect this can be quite annoying. Will work on this.

File: MANIFEST.in (hunk 4 above)
Comment (themperek) [reply MANIFEST.in]: I tried to keep it minimal that seems to work.

File: setup.py (hunk 2 above)
Comment (themperek) [reply setup.py]: Moved to cocotb/share. I have kept a symbolic link. Can remove but I think we should have this for some time not to break everyone flow. Could add some warning somewhere.

File: setup.py (hunk 3 above)
Comment (themperek) [reply setup.py]: Is VERSION used anywhere?

Not better would be something like:
VERSION=`(shell python -c "import pkg_resources; pkg_resources.get_distribution("bdaq53").version)`

**Comment by imphil** [comment]
Thanks @themperek, nice! I didn't look to closely yet (I'll do that tomorrow), just as a couple quick starters:

> I have changed cocotb-path to cocotb-config at the moment it just returns path to share folder. Shout it be rather cocotb-config --prefix or cocotb-config --makefiles ?

It would be probably good to add the arg parsing feature from the beginning, otherwise we run into backwards-compat problems again as soon as we add them.

> I have kept symbolic links to share folders. I think it is better to keep backward compatibility for some time. Could add a warning.

If users specify COCOTB=/path/to/cocotb/srcdir/cocotb/share and run their existing Makefiles, doesn't that do the trick already? So the only thing users need to do is set a different COCOTB environment variable if they update their cocotb installation. That sounds like a simple-enough change to me not to warrant a symlink. (Since you cannot be surprised by that, you actively need to update cocotb in the first place.)

> If it is OK I would add tox in a separate PR.

Sure. However it might actually help solving the current problems with CI (see below).

> Not sure why travis fails at the moment on 2.7.

The tests fail because the "cocotb" module cannot be found, so most likely the PYTHONPATH path is wrong -- which shouldn't need to be set anywhere in the first place if the package is installed properly. 
Also Dockerfile currently sets `COCOTB=xxx`, this can probably go as well.

Let me know if you get stuck, I'll have a closer look tomorrow.

**Comment by themperek** [comment]
>>I have kept symbolic links to share folders. I think it is better to keep backward compatibility for some time. Could add a warning.

>If users specify COCOTB=/path/to/cocotb/srcdir/cocotb/share and run their existing Makefiles, doesn't that do the trick already? So the only thing users need to do is set a different COCOTB environment variable if they update their cocotb installation. That sounds like a simple-enough change to me not to warrant a symlink. (Since you cannot be surprised by that, you actively need to update cocotb in the first place.)

If one has to change this then also can do `COCOTB=$(cocotb-path --prefix )` but this would need `python setup.py develop/install` or `pip install cocotb` before.

I would suggest releasing before we merge this (1.0.1?) and then have a new version (1.1?) with package support (upload to pipy) and update documentation. Let me know what you think.

File: MANIFEST.in [hunk 5]
```diff
@@ -1,10 +1,4 @@
-recursive-include bin *
-recursive-include cocotb *.py *.c Makefile
-recursive-include gpi *.c *.h Makefile
-recursive-include vpi_shim *.c Makefile
+recursive-include cocotb/share *.c* *.h Makefile.*
```
[the remark this answers is not in the document]
Comment (themperek) [reply MANIFEST.in]: Yes. I  have tried and this has to to make packaging working for now.

File: cocotb/share/makefiles/Makefile.pylib.Darwin [hunk 6]
```diff
@@ -39,6 +39,11 @@ endif
 NATIVE_ARCH=$(shell uname -m)
 ARCH?=$(NATIVE_ARCH)
 
+VERSION?=$(shell $(PYTHON_BIN) -c 'import pkg_resources; print(pkg_resources.get_distribution("cocotb").version)')
```
[the remark this answers is not in the document]
Comment (themperek) [reply cocotb/share/makefiles/Makefile.pylib.Darwin]: Yes. This is already in but this may break backward compatibility (not `cocotb-config`). This case it will be empty but should not break anything.  Depending on the decision.

**Comment by imphil** [comment]
I've tested this and the package looks good to me. Let's keep this as-is for now and not add more changes to it. The commit history is already a bit hard to follow and it would be nice to get it rebased and cleaned up into logical commits (e.g. one for each add cocotb-config, move files, use cocotb-config, and finally the travis changes). Let me know if you need help, it requires a bit of git experience.

**Comment by themperek** [comment]
>The commit history is already a bit hard to follow and it would be nice to get it rebased and cleaned up into logical commits
@imphil Not sure if I can do this "into logical commits" apart form squash all together or make a new pull request.

**Comment by eric-wieser** [comment]
Not sure how a new PR would help here @themperek, unless your goal is to start with a clean slate regarding comments.

One approach I often take to cleaning up difficult history full of iteration is:

* `git tag in-case-I-screw-up`
* `git reset --mixed 4aab2d90b4e5b7fd057b185a7b234dc87397301a`  (the commit you started at)
* Repeat until no changes remain:
   * `git add -p`, to add things piece by piece (easier through a GUI tool IMO)
   * `git commit`

**Comment by themperek** [comment]
@eric-wieser This seems to be not so straight since there was a rebase in the middle.

**Comment by imphil** [comment]
No worries, I'll do the rebase in the coming days.

**Comment by themperek** [comment]
In case someone wants to try: `pip install cocotb==1.0.dev3`

**Comment by imphil** [comment]
With the 1.1 release out now I'll look at this one again. I'll rebase this PR in the coming days and get it merged, ETA end of the week.

**Comment by Matthewar** [comment]
With [PEP517](https://www.python.org/dev/peps/pep-0517/) being [supported by pip](https://github.com/pypa/pip/pull/4144) now, should we be using this instead of setuptools?

**Comment by imphil** [comment]
I've looked through this PR and split it up and rebased it on top of the current master. I'll now step-by-step open PRs for the individual changes in the hope that this will lead to the least breakage for this rather large PR. This included also a couple small functional changes, I'll highlight them as we move along.

The current patch series is in https://github.com/imphil/cocotb/commits/packaging-rebased. Note that this series will be rebased as we iterate through the sub-PRs.

**Comment by imphil** [comment]
Starting now with #799 and #800 as first small independent fixes. More to follow as soon as these two have gone in, awaiting review by @themperek.

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    packages=find_packages(),
+    include_package_data=True,
+    package_data={'cocotb': package_files('cocotb/share')},
+    entry_points={
+        'console_scripts': [
+            'cocotb-config=cocotb.config:main',
+        ]
+    },
+    platforms='any',
+)
```
Comment (cmarqu) [inline setup.py]: I guess we should add at least the following classifiers here:
```
    classifiers=[
        "Programming Language :: Python :: 2.7",
        "Programming Language :: Python :: 3",
        "License :: OSI Approved :: BSD License",
        "Topic :: Scientific/Engineering :: Electronic Design Automation (EDA)",
    ],
```

Comment (themperek) [reply setup.py]: Good point. Can you make a PR for this? Now it is in master.
